com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -23,29 +33,154 @@
+	private Notifier[] notifiers;
+	private List<Consumer<AdvancedViatraQueryEngine>> initializationActions = new ArrayList<>();
 
-	private ViatraQueryAdapter(AdvancedViatraQueryEngine engine, Project project) {
-		this.engine = engine;
+	private ViatraQueryAdapter(Project project, Notifier[] notifiers) {
 		this.project = project;
+		this.notifiers = notifiers;
 	}
-	
+
+	/**
+	 * @deprecated Use {@link #getInitializedEngineChecked()} or {@link #getInitializedEngine()} instead
+	 * @return initialized engine
+	 * @throws ViatraQueryException if no query engine was initialized before and initialization failed
+	 */
+	@Deprecated
 	public AdvancedViatraQueryEngine getEngine() {
+		return getInitializedEngineChecked();
+	}
+
+	/**
+	 * Returns an initialized VIATRA query engine stored in the adapter.
+	 * If a query engine was initialized earlier, that engine will be returned;
+	 * if no query engine was initialized earlier, a new one is created. If
+	 * the initialization fails, e.g. the engine was requested before the
+	 * model was loaded entirely, a {@link ViatraQueryException} is thrown.
+	 * </p>
+	 * This method should only be called if the engine initialization is
+	 * expected to be completed successfully, otherwise it is recommended
+	 * to use {@link #getInitializedEngine()},
+	 * {@link #getInitializedEngine(Consumer)}
+	 * or {@link #requireMatcher(IQuerySpecification)} instead
+	 * as they are aimed to handle the initialization issues better.
+
+	 * @return the initialized engine
+	 * @throws ViatraQueryException if no query engine was initialized before and initialization failed
+	 */
+	public AdvancedViatraQueryEngine getInitializedEngineChecked() {
+		return getInitializedEngine()
+				.orElseThrow(() ->
+					new ViatraQueryException(MESSAGE_ENGINE_NOT_READY, MESSAGE_ENGINE_NOT_READY));
+	}
+
+	/**
+	 * Provide the initialized VIATRA if possible.<br />
+	 * If the engine hasn't been initialized but it can be, then it will be initialize and
+	 * every initialization action registered before will be executed.
+	 * 
+	 * @return the initialized VIATRA engine if possible, otherwise an empty Optional
+	 * @see ViatraQueryAdapter#getInitializedEngine(Consumer)
+	 * @see ViatraQueryAdapter#requireMatcher(IQuerySpecification)
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine() {
+		if(!isInitialized) {
+			initializeEngine();
+		}
 		return engine;
 	}
 	
+	/**
+	 * Checks if an engine is available and execute the given action on it if yes.
+	 * If not, the action will be registered in the list of actions which need to be done 
+	 * when the engine is ready and will be executed when the engine become available.
+	 * 
+	 * @param action the operation which would like to use the initialized VIATRA engine
+	 * @return the initialized VIATRA engine of it is available, otherwise empty Optional.
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine(Consumer<AdvancedViatraQueryEngine> action) {
```

Comment:
From user perspective, I assume we execute a non-blocking action even if the engine is not yet ready. This means the user does not really interested in whether the engine is ready or not. Hence, why would he need the return value of this method? I would suggest changing to `public void  executeActionWhenEngineInitialized(Consumer<AdvancedViatraQueryEngine> action)` without returning value.

Reply:
Good catch. We had discussed this with @tamasborbas in person as well and he will provide a newer version that handles this case (and updates the API a bit).

Reply:
I might also mark the method somehow with "async" (e.g. in the name) to make sure the users won't try to check the result of the action immediately. Or even adding a callback function to let the user know when the action is executed. But if it is against the naming convention or these imply unnecessary modification, I'm okay with the current change and with the entire PR.

Reply:
This method is strange, because it might be either sync (if the engine is already initialized) or async (if the engine is not initialized yet), so the _async postfix would be misleading.

Your original recommendation of `whenEngineInitialized` is really too long; the name `executeActionOnEngine` is the best way, as the action cannot be executed when the engine is not initialized yet; but there is only a single engine associated with each adapter, so this should be clear what is it referring to.